File: assets/assets-api/src/test/java/com/github/ykiselev/assets/SimpleAssetsTest.java
/*
 * Copyright 2017 Yuriy Kiselev (uze@yandex.ru)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.ykiselev.assets;

import com.github.ykiselev.wrap.Wraps;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.nio.channels.ReadableByteChannel;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public class SimpleAssetsTest {

    private final Resources resources = mock(Resources.class);

    private final ReadableAssets readableAssets = mock(ReadableAssets.class);

    @SuppressWarnings("unchecked")
    private final ReadableAsset<Double, Void> readableAsset = mock(ReadableAsset.class);

    private final Assets assets = new SimpleAssets(resources, readableAssets);

    @BeforeEach
    public void setUp() {
        when(resources.open(any(String.class)))
                .thenReturn(
                        Optional.of(mock(ReadableByteChannel.class))
                );
        when(readableAsset.read(any(ReadableByteChannel.class), any(Recipe.class), any()))
                .thenReturn(Wraps.simple(Math.PI));
    }

    @Test
    public void shouldLoad() {
        when(readableAssets.resolve(any(String.class), any(Recipe.class)))
                .thenReturn(readableAsset);
        when(readableAsset.read(any(ReadableByteChannel.class), any(Recipe.class), eq(assets)))
                .thenReturn(Wraps.simple(Math.PI));
        assertEquals(Math.PI, assets.load("x", DefaultRecipe.of(Double.class)).value(), 0.00001);
    }
}

File: assets/assets-api/src/test/java/com/github/ykiselev/assets/SimpleAssetsTest.kt
/*
 * Copyright 2017 Yuriy Kiselev (uze@yandex.ru)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.ykiselev.assets

import com.github.ykiselev.wrap.Wraps.noop
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import java.nio.channels.ReadableByteChannel

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
class SimpleAssetsTest {

    private val resources = mock<Resources>()
    private val readableAssets: ReadableAssets = mock<ReadableAssets>()
    private val readableAsset = mock<ReadableAsset<Double, Void>>()
    private val assets: Assets = SimpleAssets(resources, readableAssets)

    @BeforeEach
    fun setUp() {
        `when`(resources.open(any<String>()))
            .thenReturn(
                mock<ReadableByteChannel>()
            )
        `when`(readableAsset.read(any<ReadableByteChannel>(), any<Recipe<*, Double, Void>>(), any()))
            .thenReturn(noop(Math.PI))
    }

    @Test
    fun shouldLoad() {
        `when`(readableAssets.resolve(any<String>(), any<Recipe<*, Double, Void>>()))
            .thenReturn(readableAsset)
        `when`(readableAsset.read(any<ReadableByteChannel>(), any<Recipe<*, Double, Void>>(), eq(assets)))
            .thenReturn(noop(Math.PI))
        Assertions.assertEquals(
            Math.PI, assets.load(
                "x", DefaultRecipe.of(
                    Double::class.java
                )
            ).value(), 0.00001
        )
    }
}